### PR TITLE
Create API for edX installations, that dispatch statistics

### PR DESCRIPTION
**Pull request by DmytryiStriletskyi**
## Features
`olga/analytics/urls.py` provides token registration and authorization, installation statistics dispatch URLs (API.

`olga/analytics/views.py` contains API implementation:
1. `token registration` —  create and return token to edX installation.
2. `token authorization` — check if received  token from edX installation exists in Acceptor database. Else return new token (called `refreshed token`) — It just happens, when Acceptor database is empty, but edX has token from previous days/requests.
3. `receive installation statistics` via token, if token is authorized.

### CODE REVIEW

File: olga/analytics/views.py [hunk 1]
```diff
@@ -5,35 +5,85 @@
...
 from django.http import HttpResponse
 from django.views.generic import View
 from django.views.decorators.csrf import csrf_exempt
 from django.utils.decorators import method_decorator
 
 from .models import InstallationStatistics, EdxInstallation
 
+HTTP_200_OK = 200
 HTTP_201_CREATED = 201
+HTTP_401_UNAUTHORIZED = 401
```
Comment (z4y4ts) [inline olga/analytics/views.py]: Use constants from `httplib`. Or `http.client` if Acceptor is Python 3.

File: olga/analytics/views.py [hunk 2]
```diff
@@ -5,35 +5,85 @@
...
+        """
+
+        access_token = uuid.uuid4().hex
+        EdxInstallation.objects.create(access_token=access_token)
+
+        token_registration_response = json.dumps({
+            'access_token': access_token
+        })
+
+        return HttpResponse(token_registration_response, status=HTTP_201_CREATED)
```
Comment (z4y4ts) [inline olga/analytics/views.py]: Use `JsonResponse` to simplify code.

File: olga/analytics/views.py [hunk 3]
```diff
@@ -54,13 +104,13 @@ def update_students_without_country_value(active_students_amount, students_per_c
 
         return students_per_country
 
-    def create_instance_data(self, received_data, secret_token):  # pylint: disable=too-many-locals
+    def create_instance_data(self, received_data, access_token):  # pylint: disable=too-many-locals
```
Comment (z4y4ts) [inline olga/analytics/views.py]: This is probably out of the scope of this PR.
But this method must be split into smaller ones and `pylint: disable=too-many-locals` should be removed.

File: olga/analytics/views.py (hunk 1 above)
Comment (DmytryiStriletskyi) [inline olga/analytics/views.py]: Done.

File: olga/analytics/views.py (hunk 2 above)
Comment (DmytryiStriletskyi) [inline olga/analytics/views.py]: Done.

File: olga/analytics/views.py (hunk 3 above)
Comment (DmytryiStriletskyi) [inline olga/analytics/views.py]: I am sorry. `pylint: disable=too-many-locals` is a cost from refactor and mind.
I refactored this method some time ago and `pyline disable` not actually relevant for now.

File: olga/analytics/views.py [hunk 4]
```diff
@@ -2,38 +2,111 @@
...
+    def post(self, request):  # pylint: disable=unused-argument
+        """
+        Receive primary edX installation request for getting access for dispatch statistics via token.
+
+        Returns HTTP-response with status 201, that means object (installation token) was successfully created.
+        """
+        access_token = self.registry_a_token_and_return_it()
+
+        logger.info('OLGA acceptor registered edX installation with token %s', access_token)
+        return JsonResponse({'access_token': access_token}, status=httplib.CREATED)
```
Comment (z4y4ts) [inline olga/analytics/views.py]: This doesn't make sense. `access_token` is actually an `EdxInstallation` object. This is very confusing.

File: olga/analytics/views.py [hunk 5]
```diff
@@ -2,38 +2,111 @@
...
+
+
+@method_decorator(csrf_exempt, name='dispatch')
+class AccessTokenAuthorization(View):
+    """
+    Provide access token authorization functionality.
+    """
+
+    @staticmethod
+    def post(request):
```
Comment (z4y4ts) [inline olga/analytics/views.py]: Split this view into two views.

1. GET - `AuthorizeAccessToken` - To check if API consumer has access to the system.
2. POST - `RegisterEdxInstance` - To register new edX instance.

File: olga/analytics/views.py [hunk 6]
```diff
@@ -48,32 +121,66 @@ def update_students_without_country_value(active_students_amount, students_per_c
...
+        """
         students_per_country['null'] = active_students_amount - sum(students_per_country.values())
 
         return students_per_country
 
-    def create_instance_data(self, received_data, secret_token):  # pylint: disable=too-many-locals
+    def extend_statistics_to_enthusiast_level(self, received_data, installation_statistics, edx_installation_object):
+        # pylint: disable=invalid-name
+        """
+        Extend installation statistics level from `Paranoid` to `Enthusiast`.
```
Comment (z4y4ts) [inline olga/analytics/views.py]: Please elaborate on this method. At the moment it's not obvious what use cases does it cover.

File: olga/analytics/views.py [hunk 7]
```diff
@@ -48,32 +121,66 @@ def update_students_without_country_value(active_students_amount, students_per_c
...
+            edx_installation_object.latitude = enthusiast_edx_installation['latitude']
+            edx_installation_object.longitude = enthusiast_edx_installation['longitude']
+            edx_installation_object.platform_name = enthusiast_edx_installation['platform_name']
+            edx_installation_object.platform_url = enthusiast_edx_installation['platform_url']
+            edx_installation_object.save()
+
+    def create_instance_data(self, received_data, access_token):
         """
-        Method provides saving instance data as object in database.
+        Provide saving edX installation data in database.
```
Comment (z4y4ts) [inline olga/analytics/views.py]: `Provide saving edX installation data in database.` -> `Save edX installation data into a database.` If I understand you correctly.

File: olga/analytics/views.py [hunk 8]
```diff
@@ -82,55 +189,54 @@ def create_instance_data(self, received_data, secret_token):  # pylint: disable=
...
         InstallationStatistics.objects.create(edx_installation=edx_installation_object, **installation_statistics)
 
+    @staticmethod
+    def is_access_token_authorized(access_token):
+        """
+        Check if access token belongs to any EdxInstallation object.
+        """
+        try:
+            EdxInstallation.objects.get(access_token=access_token)
+            logger.info('edX installation with token %s was successfully authorized', access_token)
```
Comment (z4y4ts) [inline olga/analytics/views.py]: `logger.debug` because we don't want to spam logs with every host access.

File: olga/charts/views.py [hunk 9]
```diff
@@ -2,78 +2,59 @@
...
-
-    @staticmethod
-    def get(request):
-        """
-        Passes graph data to frontend.
-        """
-
+    try:
         first_datetime_of_update_data = InstallationStatistics.objects.first().data_created_datetime
         last_datetime_of_update_data = InstallationStatistics.objects.last().data_created_datetime
```
Comment (z4y4ts) [inline olga/charts/views.py]: Order of the returned objects is not guaranteed.
It's better to use `Min` and `Max` aggregation function here.

File: requirements_dev.txt [hunk 10]
```diff
@@ -0,0 +1,4 @@
+flake8==3.3.0
+flake8_docstrings==1.1.0
+pep8==1.7.0
+pylint==1.7.2
```
Comment (z4y4ts) [inline requirements_dev.txt]: 👍  for requirements split.

File: olga/__init__.py [hunk 11]
```diff
@@ -0,0 +1 @@
+# flake8: noqa
```
Comment (DmytryiStriletskyi) [inline olga/__init__.py]: `D104 Missing docstring in public package`.

File: olga/analytics/forms.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
+"""
+Forms for received edX installation data.
+"""
+
+from django import forms
+
+from .models import EdxInstallation, InstallationStatistics
+
+# pylint: disable-all
+# flake8: noqa
```
Comment (DmytryiStriletskyi) [inline olga/analytics/forms.py]: Forms are self documenting and easy to implement and code.
Do I need add documentation or disable just documentation errors?

File: olga/analytics/models.py [hunk 13]
```diff
@@ -33,40 +37,39 @@ class EdxInstallation(models.Model):
     Model that stores overall data received from the edx-platform.
     """
 
-    secret_token = models.CharField(max_length=255, null=True)
+    access_token = models.UUIDField(max_length=255, null=True)
```
Comment (DmytryiStriletskyi) [inline olga/analytics/models.py]: Done.

File: olga/analytics/models.py [hunk 14]
```diff
@@ -149,3 +156,111 @@ def worlds_students_per_country_statistics(cls):  # pylint: disable=invalid-name
                 world_students_per_country[country] += count
 
         return dict(world_students_per_country.items())
+
+    @staticmethod
+    def get_student_amount_percentage(country_count_in_statistics, all_active_students_in_statistics):
```
Comment (DmytryiStriletskyi) [inline olga/analytics/models.py]: Done.

File: olga/analytics/models.py [hunk 15]
```diff
@@ -149,3 +156,111 @@ def worlds_students_per_country_statistics(cls):  # pylint: disable=invalid-name
...
+        # pylint: disable=invalid-name
```
Comment (DmytryiStriletskyi) [inline olga/analytics/models.py]: Old disable from previous name, that was refactored.

File: olga/analytics/models.py [hunk 16]
```diff
@@ -149,3 +156,111 @@ def worlds_students_per_country_statistics(cls):  # pylint: disable=invalid-name
...
+            country_count_in_statistics / all_active_students_in_statistics * 100, '.2f'
+        )
+
+        if students_amount_percentage == '0.00':
+            students_amount_percentage = '~0'
+
+        return students_amount_percentage
+
+    @staticmethod
+    def is_country_exists(country):
```
Comment (DmytryiStriletskyi) [inline olga/analytics/models.py]: Done.

File: olga/analytics/models.py [hunk 17]
```diff
@@ -149,3 +156,111 @@ def worlds_students_per_country_statistics(cls):  # pylint: disable=invalid-name
...
+        """
+        Check if value is not a null.
+        """
+        if country != 'null':
+            return True
```
Comment (DmytryiStriletskyi) [inline olga/analytics/models.py]: Done.

File: olga/analytics/models.py [hunk 18]
```diff
@@ -149,3 +156,111 @@ def worlds_students_per_country_statistics(cls):  # pylint: disable=invalid-name
...
+        """
+        if not tabular_format_countries_list:
+            tabular_format_countries_list.append(('Unset', 0, 0))
+
+            # Unset is not a country
+            countries_amount = 0
+
+        else:
+            # Delete unset country point from list
+            countries_amount = len(tabular_format_countries_list) - 1
```
Comment (DmytryiStriletskyi) [inline olga/analytics/models.py]: I will improve comment above `246`. During calculation `tabular_format_countries_list` fills up `Unset` counter, that count students without country.

Actually `Unset` is not a country, we need to delete it from `countries_amount`. Do not we?

Comment (DmytryiStriletskyi) [inline olga/analytics/models.py]: I did cool refactor for that, improved logic and cleared `get_countries_amount` method. Thanks.

File: olga/analytics/models.py [hunk 19]
```diff
@@ -149,3 +156,111 @@ def worlds_students_per_country_statistics(cls):  # pylint: disable=invalid-name
...
+
+        return countries_amount
+
+    @classmethod
+    def get_worlds_students_per_country_data_to_render(cls):  # pylint: disable=invalid-name
+        """
+        Gather convenient and necessary data formats to render it from view.
+        """
+        # pylint: disable=invalid-name
+        worlds_students_per_country_country_count_accordance = \
```
Comment (DmytryiStriletskyi) [inline olga/analytics/models.py]: Done.

File: olga/analytics/utils.py [hunk 20]
```diff
@@ -1,133 +1,30 @@
...
-                        "courses_amount": courses,
-                        "data_created_datetime": dt,
-                        "edx_installation": pk,
-                        "statistics_level": 1,
-                        "students_per_country": "{\"RU\": 2632, \"CA\": 18543, \"UA\": 2011, \"null\": 1}"
-                    }
-                }
+def installation_statistics_forms_checker(receive_installation_statistics_post_method):  # pylint: disable=invalid-name
+    """
+    Checksare edX overall installation info and statistics valid with corresponding forms.
```
Comment (DmytryiStriletskyi) [inline olga/analytics/utils.py]: Rewrite docstring.

File: olga/analytics/utils.py [hunk 21]
```diff
@@ -1,133 +1,30 @@
...
 
-                instance_list.append(installation_statistics)
-                instance_list.append(edx_installation)
+    Returns HTTP-response with status 401, that means at least one of two forms is not valid.
+    """
+    def receive_installation_statistics_post_method_wrapper(request, *args, **kwargs):  # pylint: disable=invalid-name
```
Comment (DmytryiStriletskyi) [inline olga/analytics/utils.py]: Done.

File: olga/analytics/utils.py [hunk 22]
```diff
@@ -1,133 +1,30 @@
...
-                        "active_students_amount_week": 1000,
-                        "active_students_amount_month": 1000,
-                        "courses_amount": courses,
-                        "data_created_datetime": dt,
-                        "edx_installation": pk,
-                        "statistics_level": 1,
-                        "students_per_country": "{\"RU\": 2632, \"CA\": 18543, \"UA\": 2011, \"null\": 1}"
-                    }
-                }
+def installation_statistics_forms_checker(receive_installation_statistics_post_method):  # pylint: disable=invalid-name
```
Comment (DmytryiStriletskyi) [inline olga/analytics/utils.py]: Yep, it is pretty good.

File: olga/analytics/views.py (hunk 7 above)
Comment (DmytryiStriletskyi) [inline olga/analytics/views.py]: Done.

File: olga/analytics/views.py (hunk 8 above)
Comment (DmytryiStriletskyi) [inline olga/analytics/views.py]: So I need to change it in the same cases in code.

File: requirements_dev.txt (hunk 10 above)
Comment (DmytryiStriletskyi) [inline requirements_dev.txt]: Thanks.

File: olga/charts/views.py (hunk 9 above)
Comment (DmytryiStriletskyi) [inline olga/charts/views.py]: Done.

File: olga/analytics/views.py (hunk 6 above)
Comment (DmytryiStriletskyi) [inline olga/analytics/views.py]: Done.

File: olga/analytics/views.py (hunk 4 above)
Comment (DmytryiStriletskyi) [inline olga/analytics/views.py]: `access_token` is not an `EdxInstallation` object, it just token to access to save instance statistics.
During registration besides `access token` it registries `EdxInstallation` object with only `access token` field - it means default `Paranoid` level instance basic data.

Or I did not understand your comment, extend it, please.

File: olga/analytics/models.py [hunk 23]
```diff
@@ -149,3 +156,111 @@ def worlds_students_per_country_statistics(cls):  # pylint: disable=invalid-name
...
+        """
+        Check if value is not a null.
+        """
+        if country != 'null':
+            return True
+
+    @staticmethod
+    def append_country_data_to_list(country_list, country, count, student_amount_percentage=None):
+        """
+        Make data for particular visualization format. Example for `datamap` or tabular view.
```
Comment (DmytryiStriletskyi) [inline olga/analytics/models.py]: Refactored.

File: olga/analytics/models.py [hunk 24]
```diff
@@ -149,3 +156,111 @@ def worlds_students_per_country_statistics(cls):  # pylint: disable=invalid-name
...
+    @staticmethod
+    def is_country_exists(country):
+        """
+        Check if value is not a null.
+        """
+        if country != 'null':
+            return True
+
+    @staticmethod
+    def append_country_data_to_list(country_list, country, count, student_amount_percentage=None):
```
Comment (DmytryiStriletskyi) [inline olga/analytics/models.py]: Refactored.

File: olga/analytics/models.py [hunk 25]
```diff
@@ -149,3 +156,111 @@ def worlds_students_per_country_statistics(cls):  # pylint: disable=invalid-name
...
+        """
+        Make data for particular visualization format. Example for `datamap` or tabular view.
+        """
+        country_data = [str(pycountry.countries.get(alpha_2=country).alpha_3), count]
+
+        if student_amount_percentage is not None:
+            country_data += [student_amount_percentage]
+
+        country_list.append(country_data)
```
Comment (DmytryiStriletskyi) [inline olga/analytics/models.py]: Refactored.

File: olga/analytics/forms.py (hunk 12 above)
Comment (DmytryiStriletskyi) [inline olga/analytics/forms.py]: Only documentation.

File: olga/analytics/models.py [hunk 26]
```diff
@@ -110,29 +113,33 @@ def overall_counts(cls):
...
         students_count = all_unique_instances.aggregate(
-            Sum('active_students_amount_day'))['active_students_amount_day__sum']
+            Sum('active_students_amount_day')
+        )['active_students_amount_day__sum']
 
         return instances_count, courses_count, students_count
 
     @classmethod
-    def worlds_students_per_country_statistics(cls):  # pylint: disable=invalid-name
+    def get_worlds_students_per_country_country_count_accordance(cls):  # pylint: disable=invalid-name
```
Comment (DmytryiStriletskyi) [inline olga/analytics/models.py]: Thanks.

File: olga/analytics/views.py [hunk 27]
```diff
@@ -2,38 +2,111 @@
...
+from .utils import installation_statistics_forms_checker
+
+logging.basicConfig()
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
 
-HTTP_201_CREATED = 201
+stream_handler = logging.StreamHandler()
+stream_handler.setLevel(logging.INFO)
```
Comment (DmytryiStriletskyi) [inline olga/analytics/views.py]: Playin with logger costs me this unnecessary peace of code, sorry.

File: olga/analytics/views.py [hunk 28]
```diff
@@ -2,38 +2,111 @@
...
 @method_decorator(csrf_exempt, name='dispatch')
-class ReceiveData(View):
+class AccessTokenRegistration(View):
     """
-    Receives and processes data from the remote edx-platform.
+    Provide access token registration functionality.
+    """
+
+    @staticmethod
+    def registry_a_token_and_return_it():
```
Comment (DmytryiStriletskyi) [inline olga/analytics/views.py]: Done.

File: olga/analytics/views.py (hunk 5 above)
Comment (DmytryiStriletskyi) [inline olga/analytics/views.py]: Discussed.

File: olga/analytics/views.py (hunk 8 above)
Comment (DmytryiStriletskyi) [inline olga/analytics/views.py]: All cases. Discussed.

**Comment by DmytryiStriletskyi** [comment]
.

File: acceptor/settings.py [hunk 29]
```diff
@@ -23,7 +23,7 @@
 SECRET_KEY = 'vpzj&s*%wyjs7u!_g1j!5az_m3^u^kj(%p_zkayd&o@0(&w@2k'
 
 # SECURITY WARNING: don't run with debug turned on in production!
-DEBUG = False
+DEBUG = True
```
Comment (DmytryiStriletskyi) [inline acceptor/settings.py]: Ok. Done.

File: olga/analytics/views.py [hunk 30]
```diff
@@ -127,8 +122,50 @@ def update_students_without_country_value(active_students_amount, students_per_c
...
+        students_per_country_updated = self.update_students_with_no_country(
+            active_students_amount_day, students_per_country_decoded
+        )
+
+        students_per_country_encoded = json.dumps(students_per_country_updated)
+
+        return students_per_country_encoded
+
+    @staticmethod
+    def does_edx_installation_extend_level_first_time(edx_installation_object):  # pylint: disable=invalid-name
```
Comment (z4y4ts) [inline olga/analytics/views.py]: This should be an `EdxInstallation`'s method, rather than view's.

File: olga/analytics/views.py [hunk 31]
```diff
@@ -127,8 +122,50 @@ def update_students_without_country_value(active_students_amount, students_per_c
...
+        If platform url exists It means edx installation already has overall information about itself.
+        Returns False and do nothing.
+
+        If platform url does not exist It means edx installation extended statistics level first time.
+        Returns True and go to update edx installation`s overall information.
+        """
+        return not edx_installation_object.platform_url
+
+    @staticmethod
+    def update_edx_instance_info(edx_installation_object, enthusiast_edx_installation):
```
Comment (z4y4ts) [inline olga/analytics/views.py]: This should be an `EdxInstallation`'s method, rather than view's.

File: olga/analytics/views.py [hunk 32]
```diff
@@ -127,8 +122,50 @@ def update_students_without_country_value(active_students_amount, students_per_c
 
         return students_per_country
```
Comment (z4y4ts) [inline olga/analytics/views.py]: Don't modify input data. Please change.

File: olga/analytics/views.py [hunk 33]
```diff
@@ -127,8 +122,50 @@ def update_students_without_country_value(active_students_amount, students_per_c
...
 
-    def extend_statistics_to_enthusiast_level(self, received_data, installation_statistics, edx_installation_object):
-        # pylint: disable=invalid-name
+    def get_students_per_country(self, students_per_country, active_students_amount_day):
+        """
+        Get students per country in json-based encoded string for saving to database.
+
+        First, it load json-based students per country string and go to calculate students without country value.
+        Second, it dumps it again to storing in convenient type, easy json loads string.
```
Comment (z4y4ts) [inline olga/analytics/views.py]: This sentence makes no sense in English.

File: olga/analytics/views.py [hunk 34]
```diff
@@ -127,8 +122,50 @@ def update_students_without_country_value(active_students_amount, students_per_c
...
+        )
+
+        students_per_country_encoded = json.dumps(students_per_country_updated)
+
+        return students_per_country_encoded
+
+    @staticmethod
+    def does_edx_installation_extend_level_first_time(edx_installation_object):  # pylint: disable=invalid-name
+        """
+        Check does edx installation extend statistics level first time.
```
Comment (z4y4ts) [inline olga/analytics/views.py]: `Check if edx installation extends statistics...`

File: olga/analytics/views.py [hunk 35]
```diff
@@ -127,8 +122,50 @@ def update_students_without_country_value(active_students_amount, students_per_c
...
+
+        If platform url exists It means edx installation already has overall information about itself.
+        Returns False and do nothing.
+
+        If platform url does not exist It means edx installation extended statistics level first time.
+        Returns True and go to update edx installation`s overall information.
```
Comment (z4y4ts) [inline olga/analytics/views.py]: `

File: olga/analytics/models.py [hunk 36]
```diff
@@ -149,3 +155,98 @@ def worlds_students_per_country_statistics(cls):  # pylint: disable=invalid-name
...
+        """
+        Calculate countries amount in worlds students per country statistics as table.
+
+        If no countries - appends `Unset` field with zero countries, else get.
+        Actually `Unset` field is not a country, so it does not fill up in countries amount.
+        """
+        if not tabular_format_countries_list:
+            return 0
+
+        return len(tabular_format_countries_list) - 1
```
Comment (z4y4ts) [inline olga/analytics/models.py]: I still don't understand why `- 1`?

File: olga/analytics/forms.py [hunk 37]
```diff
@@ -6,16 +6,18 @@
 
 from .models import EdxInstallation, InstallationStatistics
 
-# pylint: disable-all
-# flake8: noqa
+# Pylint does not has opportunity to disable duplicate code (forms and admin has the same peaces).
```
Comment (DmytryiStriletskyi) [inline olga/analytics/forms.py]: Done.

File: olga/analytics/views.py (hunk 33 above)
Comment (DmytryiStriletskyi) [inline olga/analytics/views.py]: Done.

File: olga/analytics/views.py (hunk 35 above)
Comment (DmytryiStriletskyi) [inline olga/analytics/views.py]: Done.

File: olga/analytics/views.py (hunk 34 above)
Comment (DmytryiStriletskyi) [inline olga/analytics/views.py]: Done.

File: olga/analytics/models.py (hunk 36 above)
Comment (DmytryiStriletskyi) [inline olga/analytics/models.py]: From method docstring: `Actually 'Unset' field is not a country, so it does not fill up in countries amount.`. I'll extend documentation.

File: olga/analytics/views.py (hunk 30 above)
Comment (DmytryiStriletskyi) [inline olga/analytics/views.py]: I agree with you. Done.

File: olga/analytics/views.py (hunk 31 above)
Comment (DmytryiStriletskyi) [inline olga/analytics/views.py]: I agree with you. Done.

File: olga/analytics/views.py [hunk 38]
```diff
@@ -139,32 +148,6 @@ def get_students_per_country(self, students_per_country, active_students_amount_
 
         return students_per_country_encoded
 
-    @staticmethod
-    def does_edx_installation_extend_level_first_time(edx_installation_object):  # pylint: disable=invalid-name
```
Comment (z4y4ts) [inline olga/analytics/views.py]: ```def does_edx_installation_extend_level_first_time(self):```

File: olga/analytics/views.py [hunk 39]
```diff
@@ -191,8 +174,8 @@ def extend_stats_to_enthusiast(self, received_data, installation_statistics, edx
 
         installation_statistics.update(enthusiast_statistics)
 
-        if self.does_edx_installation_extend_level_first_time(edx_installation_object):
-            self.update_edx_instance_info(edx_installation_object, enthusiast_edx_installation)
+        if EdxInstallation().does_edx_installation_extend_level_first_time(edx_installation_object):
```
Comment (z4y4ts) [inline olga/analytics/views.py]: ```edx_installation_object.does_edx_installation_extend_level_first_time()```

File: olga/analytics/models.py [hunk 40]
```diff
@@ -149,3 +181,101 @@ def worlds_students_per_country_statistics(cls):  # pylint: disable=invalid-name
...
+        Tabular format countries list can be empty - countries amount is zero.
+        Tabular format countries list can be not empty - it contains particular country-count accordance
+        and `Unset` field, that has students without country amount.
+
+        Actually `Unset` field is not a country, so it does not fill up in countries amount.
+        """
+        if not tabular_format_countries_list:
+            return 0
+
+        return len(tabular_format_countries_list) - 1
```
Comment (z4y4ts) [inline olga/analytics/models.py]: # Exclude 'Unset' country from count

Comment (DmytryiStriletskyi) [inline olga/analytics/models.py]: Done.

**Pull request merged**